File: core/src/test/java/io/whyjvm/agent/DifferentialTest.java
package io.whyjvm.agent;

import io.whyjvm.capture.Dimensions;
import io.whyjvm.capture.IncidentRecord;
import io.whyjvm.capture.ThreadActivity;
import io.whyjvm.capture.TriageSignals;
import io.whyjvm.trigger.IncidentType;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class DifferentialTest {

    private static IncidentRecord slow(long durationMs, TriageSignals sig, ThreadActivity ta) {
        return IncidentRecord.initial(
                        "inc", Instant.now(), "GET /x", IncidentType.SLOW, "fp",
                        ta != null ? ta.thread() : null, durationMs, 1, null, null, null)
                .withEvidence(sig, new Dimensions(null, null, null, ta, null), null);
    }

    @Test
    void externalWaitRulesOutJvmNoise() {
        // Thread dormiu 10s; alocacao alta (143MB) e JVM-wide ruido.
        IncidentRecord rec = slow(10005,
                new TriageSignals(3, 4, 8, 0, 143L * 1024 * 1024),
                new ThreadActivity("http-nio-1", 10000, null, 0, 0, 0, 0));
        String d = String.join(" | ", Differential.of(rec));

        assertTrue(d.contains("Pausas de GC"), d);
        assertTrue(d.contains("Contencao de lock"), d);
        assertTrue(d.contains("CPU: a thread"), d);
        assertTrue(d.contains("ruido de fundo"), d);
        // A espera externa NAO e descartada — e a causa.
        assertFalse(d.contains("Espera externa"), d);
    }

    @Test
    void keepsLiveCandidates() {
        // GC 812ms + thread em CPU (38 amostras) + alocacao alta.
        IncidentRecord rec = slow(4200,
                new TriageSignals(3, 812, 1340, 0, 1288490188L),
                new ThreadActivity("http-nio-3", 0, null, 0, 0, 0, 38));
        String d = String.join(" | ", Differential.of(rec));

        assertTrue(d.contains("Contencao de lock"), d);
        assertTrue(d.contains("Espera externa"), d);
        // Candidatos vivos NAO descartados: GC, CPU, alocacao.
        assertFalse(d.contains("Pausas de GC"), d);
        assertFalse(d.contains("CPU: a thread"), d);
        assertFalse(d.contains("Pressao de alocacao"), d);
    }

    @Test
    void noSignalsRulesOutNothing() {
        IncidentRecord rec = IncidentRecord.initial(
                "inc", Instant.now(), "GET /x", IncidentType.SLOW, "fp", "t", 100, 1, null, null, null);
        assertTrue(Differential.of(rec).isEmpty());
    }
}

File: analysis-service/internal/agent/differential_test.go
package agent

import (
	"strings"
	"testing"

	"github.com/whyjvm/analysis-service/internal/incident"
)

func joinDiff(d []string) string { return strings.Join(d, " | ") }

func TestDifferentialExternalWaitRulesOutJvmNoise(t *testing.T) {
	// db-search?slow=true: thread dormiu 10s; alocacao alta (143MB) e JVM-wide ruido.
	rec := &incident.Record{
		DurationMs: 10005,
		TriageSignals: &incident.TriageSignals{
			GcCount: 3, LongestGcPauseMs: 4, TotalGcPauseMs: 8,
			TotalLockWaitMs: 0, TotalAllocBytes: 143 * 1024 * 1024,
		},
		Dimensions: incident.Dimensions{
			ThreadActivity: &incident.ThreadActivity{Thread: "http-nio-1", SleepMs: 10000, CpuSamples: 0},
		},
	}
	d := joinDiff(differential(rec))

	for _, want := range []string{"Pausas de GC", "Contencao de lock", "CPU: a thread", "ruido de fundo"} {
		if !strings.Contains(d, want) {
			t.Fatalf("esperava descartar %q em:\n%s", want, d)
		}
	}
	// A espera externa NAO e descartada — ela e a causa.
	if strings.Contains(d, "Espera externa") {
		t.Fatalf("nao deveria descartar espera externa (e a causa):\n%s", d)
	}
}

func TestDifferentialKeepsLiveCandidates(t *testing.T) {
	// slow-gc: GC 812ms + thread em CPU (38 amostras) + alocacao alta.
	rec := &incident.Record{
		DurationMs: 4200,
		TriageSignals: &incident.TriageSignals{
			GcCount: 3, LongestGcPauseMs: 812, TotalGcPauseMs: 1340,
			TotalLockWaitMs: 0, TotalAllocBytes: 1288490188,
		},
		Dimensions: incident.Dimensions{
			ThreadActivity: &incident.ThreadActivity{Thread: "http-nio-3", CpuSamples: 38},
		},
	}
	d := joinDiff(differential(rec))

	// Descartados: lock e espera externa.
	if !strings.Contains(d, "Contencao de lock") || !strings.Contains(d, "Espera externa") {
		t.Fatalf("esperava descartar lock e espera externa:\n%s", d)
	}
	// Candidatos vivos NAO descartados: GC, CPU, alocacao.
	for _, live := range []string{"Pausas de GC", "CPU: a thread", "Pressao de alocacao"} {
		if strings.Contains(d, live) {
			t.Fatalf("nao deveria descartar candidato vivo %q:\n%s", live, d)
		}
	}
}

func TestDifferentialNilWithoutSignals(t *testing.T) {
	if d := differential(&incident.Record{DurationMs: 100}); d != nil {
		t.Fatalf("sem triageSignals nao deveria descartar nada, veio %v", d)
	}
}
